src/Manifolds.jl

```diff
@@ -102,6 +106,94 @@ function hat(M::Manifold, x, vⁱ)
+The general formula to compute the minimizer reads
+````math
+\argmin_{y\in\mathcal M} \frac{1}{2}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}^2(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemian [`distance`](@ref).
+
+Finally `stop_tol` denotes the minimal change between to iterates then to stop
+and `stop_iter` the maximal number of iterations. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in 
+> Afsari, B; Tron, R.; Vidal, R.: On the Convergence of Gradient
+> Descent for Finding the Riemannian Center of Mass,
+> SIAM Journal on Control and Optimization (2013), 51(3), pp. 2230–2260, 
+> doi: [10.1137/12086282X](https://doi.org/10.1137/12086282X),
+> arxiv: [1201.0925](https://arxiv.org/abs/1201.0925">1201.0925)
+"""
+function mean(M::Manifold, x::Vector{T};
+    weights=1/length(x)*ones(length(x)),
+    stop_atol=10^-7,
+    stop_iter=100
+    ) where {T}
+    iter = 0
+    y=x[1]
+    yOld = y
+    while ( ( iter==0 || distance(M,y,yOld) > stop_atol ) && (iter < stop_iter) )
+        iter += 1
+        yOld = y
+        y = exp(M,yOld,  1/2*sum( weights.*log.(Ref(M), Ref(yOld),x) )  )
+    end
+    return y
+end
+
+@doc doc"""
+    median(M,x; weights=1/n*ones(n), stop_tol=10^-10, stop_iter=10000, use_rand = false )
+
+computes the Riemannian median of the vector `x`  of `n` points on the
+[`Manifold`](@ref) `M`. This function is nonsmooth (i.e nondifferentiable) and
+uses a cyclic procimal point scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian median. The general formula to compute
+the minimizer reads
+````math
+\argmin_{y\in\mathcal M}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemian [`distance`](@ref).
+
+Finally `stop_tol` denotes the minimal change between to iterates then to stop
+and `stop_iter` the maximal number of iterations, and the cycle order of the
+cyclic proximal point method can be set to `random`, which sometimes performes
+faster.  For more stopping criterai and details check the
+[`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in Algorithm 4.3 and 4.4 in 
+> Bačák, M: Computing Medians and Means in Hadamard Spaces.
+> SIAM Journal on Optimization (2014), 24(3), pp. 1542–1566,
+> doi: [10.1137/140953393](https://doi.org/10.1137/140953393),
+> arxiv: [1210.2145](https://arxiv.org/abs/1210.2145)
+"""
+function median(M::Manifold, x::Vector{T};
+    weights=1/length(x)*ones(length(x)),
+    stop_atol=10^-10,
+    stop_iter=10000,
+    use_random = false
+    ) where {T}
+    n = length(x)
+    y=x[1]
+    yOld = y
+    order = 1:n
+    iter = 0
+    while ( ( iter==0 || distance(M,y,yOld) > stop_atol ) && (iter < stop_iter) )
+        iter += 1
+        λ = 1/(iter+1)
+        yOld = y
+        order = use_random ? randperm(n) : order
```

Comment:
That's almost surely a type instability. Making `use_random` a `::Val` or initializing `order` by `collect(1:n)` would fix it.

Reply:
I think it's easier to switch to `shuffle!` then.